### PR TITLE
URO-208: orcidlink main view

### PR DESCRIPTION
**Pull request by eapearson**
Populates the initial view for orcidlink based on whether the user is linked or not. 

This set of changes adds improved orcidlink service api support, displays orcid link and orcid profile summary information in the initial view, for linked users, and a call to action for unlinked users.

This set of changes is about 600 LOC without tests, 1200 with tests. It could be pared down further, but this minimizes the changes without losing core functionality.

This view is incomplete, as the full set of changes to implement this view would be about 1200 LOC. I trimmed out a lot of content, such as help,  some stub handling of button events, support for developer and manager detection and ui support. These changes have been made and are ready to layer on top of these changes once they are merged in.

One dependency was added - material icons. They seem to be required for integration into some MUI components, such as the Accordion.

### CODE REVIEW

File: src/features/orcidlink/HomeUnlinked.tsx [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        <Card variant="outlined">
+          <CardHeader title="Create Your KBase ORCID® Link!" />
+          {/* 
+            A note on the padding and margin overrides. Mui is not very smart about components
+            embedded in card content. The card content has a padding on all subelements (header, 
+            content, actions) which is added to any margins and padding of components contained 
+            within, so it pretty easily blows up into a lot of blank space. 
+            To keep the empty space under control (and others have commented on this 
+            in GH issues), one must fiddle with padding and margins on a case-by-case
+            basis.
```
Comment (dauglyon) [inline src/features/orcidlink/HomeUnlinked.tsx:29]: Seems fine for now! @codytodonnell, do you have recommendations here? You are more familiar than I...

**Review by dauglyon — COMMENTED** [review]
Looking good, most of the comments I left are pretty minor, but I had some questions. Feel free to poke me for re-review or if you want to go over anything on a call.

File: src/features/orcidlink/HomeUnlinked.tsx (hunk 1 above)
Comment (codytodonnell) [reply src/features/orcidlink/HomeUnlinked.tsx:29]: It sounds like you're bumping against some known issues so overrides make sense. If you have to repeat the overrides you could either create a wrapper component or make a global `styleOverride` for the `Card` component in the MUI theme object. 

Another possibility could be to wrap the inner `CardContent` in a `Stack`. I'm not sure but that could offset the subelement padding being added by Card. It may also help to omit using the `CardHeader` and just include the title in the `CardContent`.

File: src/features/orcidlink/Home/index.tsx [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import Home from './Home';
+
+export default function HomeController() {
+  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+  const username = useAppSelector(authUsername)!;
+
+  usePageTitle('KBase ORCID Link');
+
+  const { data, error, isError, isSuccess } =
+    orcidlinkAPI.useOrcidlinkInitialStateQuery({ username });
```
Comment (dauglyon) [inline src/features/orcidlink/Home/index.tsx:15]: To skip the query when username is undefined
```suggestion
  import { skipToken } from '@reduxjs/toolkit/query';
  ...
  const { data, error, isError, isSuccess } =
    orcidlinkAPI.useOrcidlinkInitialStateQuery(username?{ username }:skipToken);
```

File: src/features/orcidlink/Home/index.tsx [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import { orcidlinkAPI } from '../../../common/api/orcidlinkAPI';
+import { useAppSelector } from '../../../common/hooks';
+import { authUsername } from '../../auth/authSlice';
+import { usePageTitle } from '../../layout/layoutSlice';
+import ErrorMessage from '../common/ErrorMessage';
+import Home from './Home';
+
+export default function HomeController() {
+  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+  const username = useAppSelector(authUsername)!;
```
[the remark this answers is not in the document]
Comment (dauglyon) [reply src/features/orcidlink/Home/index.tsx:10]: Alternately, you could just let the type of username remain `| undefined` and use `skipToken` to conditionally call the query. Its usually better to conditionally render the component in the parent, though.

[the remark this answers is not in the document]
Comment (dauglyon) [reply src/features/orcidlink/Home/index.tsx:10]: see comment on line 14

File: src/common/api/orcidlinkAPI.ts [hunk 4]
```diff
@@ -1,56 +1,79 @@
 import { baseApi } from '.';
-import { jsonRpcService } from './utils/serviceHelpers';
+import { LinkRecordPublic, ORCIDProfile } from './orcidLinkCommon';
+import { jsonRpc2Service } from './utils/serviceHelpers';
 
-// orcidlink system types
+// is-linked
+
+export type IsLinkedResult = boolean;
```
[the remark this answers is not in the document]
Comment (eapearson) [reply src/common/api/orcidlinkAPI.ts:7]: I often use this style because to me It helps keep the api implementation consistent - one expects a parameter type and a result type named in a specific manner. In fact, I typically create the api method, with named types, and then back fill the type definitions. I tend not to go back and remove a type like this if it resolves down to a simple type alias.
That said, I know that some prefer not to use a type alias like this; I can easily change it.

File: src/common/api/utils/common.ts [hunk 5]
```diff
@@ -22,6 +22,7 @@ type JsonRpcError = {
   };
 };
 
+// https://github.com/reduxjs/redux-toolkit/blob/7cd8142f096855eb7cd03fb54c149ebfdc7dd084/packages/toolkit/src/query/fetchBaseQuery.ts#L48
```
[the remark this answers is not in the document]
Comment (eapearson) [reply src/common/api/utils/common.ts:25]: I'll just remove it. Didn't mean to leave it in there, it was just a mental note for easy reference. When using VSC or another editor with lookup/introspection capabilities, one can jump to the definition, which will be accurate (the link will become outdated quickly, if not already!)

File: src/features/orcidlink/common/Scopes.tsx [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+      return (
+        <Alert severity="error" title="Not a valid scope" key={scope}>
+          Invalid scope: {scope}
+        </Alert>
+      );
+    }
+    const { orcid, help, seeAlso } = getScopeHelp(scope);
+    return (
+      <Accordion key={scope}>
+        <AccordionSummary expandIcon={<ExpandMore />}>
```
[the remark this answers is not in the document]
Comment (eapearson) [reply src/features/orcidlink/common/Scopes.tsx:44]: Right, same as the above comment. Also need to use the style override for the custom rotation.

File: src/features/orcidlink/images.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
+/**
+ * Constructs urls for images used in the orcidlink ui
+ *
+ * Note that the usual technique, importing the image, which creates a url back
+ * into the web app, does not work with the current state of Europa dependencies.
+ * Therefore, we use hardcoded image paths, combined with the PUBLIC_URL.
+ * When Europa deps are updated, or it is switched to vite, the image imports
+ * should be re-enabled.
+ */
```
[the remark this answers is not in the document]
Comment (eapearson) [reply src/features/orcidlink/images.ts:9]: This relates to the error:

```
ERROR in ./src/features/orcidlink/assets/ORCID-iD_icon-vector.svg
Module build failed (from ./node_modules/file-loader/dist/cjs.js):
Error: error:0308010C:digital envelope routines::unsupported
    at new Hash (node:internal/crypto/hash:68:19)
    at Object.createHash (node:crypto:138:10)
    at getHashDigest (/workspace/node_modules/loader-utils/lib/getHashDigest.js:46:34)
    at /workspace/node_modules/loader-utils/lib/interpolateName.js:113:11
    at String.replace (<anonymous>)
    at interpolateName (/workspace/node_modules/loader-utils/lib/interpolateName.js:110:8)
    at Object.loader (/workspace/node_modules/file-loader/dist/index.js:29:48)
```

which occurs when trying to import the image directly. This is related to the libraries used to load the image and security restrictions in node > 16. Some have said it was fixed by CRA updates (and the transitive dependency upon webpack), others say not. There are other workarounds, but they are insecure.

File: src/features/orcidlink/utils.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
+/**
+ * Creates a URL object given the configured KBase domain and a path.
+ *
+ * This is suitable for creating external links in the same KBase deploy environment.
+ *
+ * @param path
+ * @returns
+ */
+export function uiURL(path: string) {
```
[the remark this answers is not in the document]
Comment (eapearson) [reply src/features/orcidlink/utils.ts:9]: Okay, I'll remove it. This is actually part of the larger set of changes that was inadvertently left in this reduced set of changes. It is intended for forming urls to be used to open external windows to any ui endpoint in the current Kbase deploy environment. I think the impetus was links to external ui endpoints while operating on localhost, but even if so we'll see what shakes out later.

File: src/features/orcidlink/common/misc.tsx [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+  );
+}
+
+export function renderLoadingOverlay(open: boolean) {
+  return (
+    <Modal open={open} disableAutoFocus={true}>
+      {renderLoading('Loading...', 'Loading ORCID Link')}
+    </Modal>
+  );
+}
```
[the remark this answers is not in the document]
Comment (eapearson) [reply src/features/orcidlink/common/misc.tsx:95]: I can have a crack at modifying Loader.

File: src/features/orcidlink/Home/index.tsx (hunk 2 above)
Comment (eapearson) [reply src/features/orcidlink/Home/index.tsx:15]: Yes, but  username is never undefined! This component is only invoked when authenticated, so I thought it cleaner to make that assumption and simplify the code.

The code pathway ensures that the orcidlink feature is only ever mounted when the app is authenticated. Yet the app state overall includes being unauthenticated.

It simplifies the implementation to assume that the user is known. For instance, the orcidlink initial state does not make sense without the authenticated username. Even though the query could be skipped if username is undefined, the query would still have to handle the case of the username being undefined.

In other words, a bunch of bother if we let username be undefined, and it seems cleaner to just assert at the beginning that it will never be undefined.

File: src/features/orcidlink/images.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (dauglyon) [reply src/features/orcidlink/images.ts:9]: Agreed that we should move to vite. Especially with this issue you note about asset loading, that could eventually be a true pain. I opened a PR a while back with an initial attempt. This may be a good thing to bring up as future work to prioritize in the UI meeting.

File: src/features/orcidlink/HomeUnlinked.tsx (hunk 1 above)
Comment (eapearson) [reply src/features/orcidlink/HomeUnlinked.tsx:29]: This may be partly addressed in MUI v6: https://github.com/mui/material-ui/pull/32471
I've made some adjustments, after which it seems to make more sense to make overrides solely to the content to remove padding and/or margin from the top and bottom of the content

File: src/features/orcidlink/Home/index.tsx (hunk 2 above)
Comment (dauglyon) [reply src/features/orcidlink/Home/index.tsx:15]: In that case maybe a `const usernameString = username ?? ''` would suffice? I'm very hesitant about assertions, as they can have really weird downstream effects, though probably not in this case. Just worried about precedent.

File: src/features/orcidlink/common/misc.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (eapearson) [reply src/features/orcidlink/common/misc.tsx:95]: In the end I left this alone -- seems not worth it, for now at least, for such a low amount of code?

File: src/common/api/orcidlinkAPI.ts [hunk 10]
```diff
@@ -1,56 +1,79 @@
...
+
+// owner-link
+
+export type OwnerLinkResult = LinkRecordPublic;
```
[the remark this answers is not in the document]
Comment (eapearson) [reply src/common/api/orcidlinkAPI.ts:11]: Well, this one wasn't even used. I think another case of a remnant left over after code was removed. (In a previous iteration there was an endpoint for just the owner-link method.

File: src/features/orcidlink/Home/index.tsx (hunk 2 above)
Comment (eapearson) [reply src/features/orcidlink/Home/index.tsx:15]: How about:

```
  // look, no non-null assertion.
  const username = useAppSelector(authUsername);

  // this condition will never occur in real life; we can, however, test it.
  if (typeof username === 'undefined') {
    throw new Error('Impossible: username is not defined');
  }
```

I've used two approaches in this situation, other than a simple non-null assertion:

- if the value is whatever-is-impossible-but-TS-can't-know-this, throw an error with the word "impossible" somewhere in the error message; the exception will never be invoked in reality, and if it could be it would handled like any other uncaught runtime error by the error boundary; that is, the invalid value would not be propagated and do unexpected things.

- pass the auth state from the authentication wrapper component as a prop to the sub-component

Comment (eapearson) [reply src/features/orcidlink/Home/index.tsx:18]: So I added the detection of undefined, throwing an exception, and a test to match the display of the message in the error boundary.

File: package-lock.json
Comment (dauglyon) [inline package-lock.json:1]: You still have changes you your package-lock related to mui/icons that should probably be removed

File: src/features/orcidlink/common/misc.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (dauglyon) [reply src/features/orcidlink/common/misc.tsx:95]: Ok sounds reasonable

**Review by dauglyon — COMMENTED** [review]
Resolved the resolved, that all looks good. However, still a few lingering comments from the previous review.

File: package-lock.json
Comment (eapearson) [reply package-lock.json:1]: Thanks for catching that. I had it lined up, but have a number of files that can't be pushed up and it got lost in the shuffle.

File: src/common/api/orcidlinkAPI.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (eapearson) [reply src/common/api/orcidlinkAPI.ts]: I did change this

File: src/features/orcidlink/HomeUnlinked.tsx (hunk 1 above)
Comment (eapearson) [reply src/features/orcidlink/HomeUnlinked.tsx]: I ended up just removing the removal of padding.

**Review by dauglyon — COMMENTED** [review]
there are still 8 unresolved comments, are they showing up for you?

**Comment by eapearson** [comment]
> there are still 8 unresolved comments, are they showing up for you?

Okay, now I see them. There were collapsed down together and not displayed.

File: src/common/api/orcidlinkAPI.ts [hunk 11]
```diff
@@ -1,56 +1,79 @@
...
-export type OwnerLinkResult = LinkRecordPublic;
+export interface ORCIDLinkLinkedUserInfoParams {
+  username: string;
+}
 
 // It is mostly a JSONRPC 2.0 service, although the oauth flow is rest-ish.
-const orcidlinkService = jsonRpcService({
+const orcidlinkService = jsonRpc2Service({
   url: '/services/orcidlink/api/v1',
   version: '2.0',
```
[the remark this answers is not in the document]
Comment (eapearson) [reply src/common/api/orcidlinkAPI.ts:66]: There is only one version for a JSON-RPC 2 api, and only one version for a JSON-RPC 1.1 api. Maybe I'm not understanding the question?

File: src/common/api/orcidLinkCommon.ts [hunk 12]
```diff
@@ -0,0 +1,73 @@
+export interface ORCIDAuthPublic {
```
[the remark this answers is not in the document]
Comment (eapearson) [reply src/common/api/orcidLinkCommon.ts:1]: Well, values of this type are used, although in the code present now, they are part of the link record.  Intellisense works if they are not expored. For an API, I tend keep the accessible interfaces public, to facilitate discovery and usage.

File: src/common/api/utils/kbaseBaseQuery.ts [hunk 13]
```diff
@@ -25,6 +32,7 @@ export interface JsonRpcQueryArgs {
   method: string;
   params?: unknown;
   fetchArgs?: FetchArgs;
+  responseHandler?: ResponseHandler;
```
[the remark this answers is not in the document]
Comment (eapearson) [reply src/common/api/utils/kbaseBaseQuery.ts:35]: It was there for a reason, but either the code using it went away or it is a future change. I'll just remove it.

File: src/common/api/utils/serviceHelpers.ts [hunk 14]
```diff
@@ -13,6 +13,18 @@ export const jsonRpcService = (
   });
 };
 
+export const jsonRpc2Service = (
```
[the remark this answers is not in the document]
Comment (eapearson) [reply src/common/api/utils/serviceHelpers.ts:16]: The two functions are currently the same, but I wanted the call to set up the orcidlink api to be a bespoke JSON-RPC 2.  But I backed out most of that code to keep the diffs smaller, and this remained to be developed later. I'll just remove it.

File: src/features/orcidlink/Home/index.tsx [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+
+  const { data, error, isError, isSuccess } =
+    orcidlinkAPI.useOrcidlinkInitialStateQuery({ username });
+
+  if (isError) {
+    return <ErrorMessage error={error} />;
+  } else if (isSuccess) {
+    return <Home isLinked={data.isLinked} info={data.info} />;
+  }
+  return null;
```
[the remark this answers is not in the document]
Comment (eapearson) [reply src/features/orcidlink/Home/index.tsx:25]: Fine, changed. I only use the react fragment on occasion to group components, and null to signal nothing to render, but it is a crazy wonderful diverse world.

File: src/features/orcidlink/HomeLinked/LinkInfo.tsx [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+      </a>
+    );
+  }
+
+  return (
+    <div>
+      <div className={styles['prop-table']}>
+        <div>
+          <div>ORCID iD</div>
+          <div>{renderORCIDId()}</div>
```
[the remark this answers is not in the document]
Comment (eapearson) [reply src/features/orcidlink/HomeLinked/LinkInfo.tsx]: Indeed, could. I just usually break out a big stretch of markup into rendering functions. They are easier to move around and share, but yeah, in this case it reduces down to just single statement.
Or does it?
Okay, so I just moved it into the misc.tsx components collection.

File: src/features/orcidlink/common/misc.tsx
[the remark this answers is not in the document]
Comment (eapearson) [reply src/features/orcidlink/common/misc.tsx:1]: Yeah, I was trying to keep the auxiliary components lightweight render functions. They are each now standalone components.

**Review by dauglyon — APPROVED** [review]
Looks good! Approved and good to merge when you're ready
